drivers/spi/spi_nrfx_spim.c

```diff
@@ -143,14 +146,52 @@ static int configure(const struct device *dev,
 	ctx->config = spi_cfg;
 	spi_context_cs_configure(ctx);
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP)
+	/* On nRF5340, the 32 Mbps speed is supported by the application core
+	 * when it is running at 128 MHz (see the Timing specifications section
+	 * in the nRF5340 PS).
+	 */
+	if (max_freq > 16000000 &&
+	    nrf_clock_hfclk_div_get(NRF_CLOCK) != NRF_CLOCK_HFCLK_DIV_1) {
+		max_freq = 16000000;
+	}
+#endif
 	/* Limit the frequency to that supported by the SPIM instance */
-	freq = get_nrf_spim_frequency(MIN(spi_cfg->frequency,
-					  get_dev_config(dev)->max_freq));
+	spim_frequency = get_nrf_spim_frequency(MIN(spi_cfg->frequency,
+						    max_freq));
+
+#if defined(CONFIG_SOC_NRF5340_CPUAPP)
+	/* For 32 Mbps speed on nRF5340, drive configuration H0H1 must be used
+	 * (see the Pin assignments section in the nRF5340 PS).
+	 */
+	nrf_gpio_pin_drive_t drive = (spim_frequency == NRF_SPIM_FREQ_32M)
```

Comment:
Just realised that the HAL drivers already perform this configuration in `nrfx_spim_init`

Reply:
I'm sorry that it took me so long to get back to this one. I reworked the PR to take advantage of the pin configuration performed by the nrfx driver. Please take a look once again.